### PR TITLE
Transaction verifier structure

### PR DESCRIPTION
**Pull request by hdevalence**
Supersedes #1100, as described here:  https://github.com/ZcashFoundation/zebra/pull/1100#issuecomment-709635240

Rather than force-pushing the rebased branch on top of that one, I created a new PR (this one) to allow comparison.

This PR:

- fills out the `transaction::Verifier` with enough structure to slot in the rest of the checks one-by-one;
- adds some helper methods to `zebra-chain` to support future checks;
- demonstrates async checks using the batch redjubjub code;
- fills in the `script::Verifier`, which does plumbing to asynchronously connect the state with `zebra-script`;
- connects the `script`, `transaction`, and `block` verifiers.

The extra checks added in #1100 are removed and will come in a follow-on PR. This way, this PR focuses on the structure and the later one on the consensus rules.

### CODE REVIEW

**Comment by hdevalence** [comment]
One issue that we should fix is the duplication in error types.  The `zebra_consensus::error` module already has error types that are supposed to exhaustively enumerate consensus rules (#1029), but these are currently duplicated by the new transaction code.

**Comment by hdevalence** [comment]
Hmm, it looks like the error type duplication is pre-existing to this PR: there's `VerifyBlockError` and `VerifyTransactionError` that both shouldn't exist. We could either fix it now, or in a later PR.  The downside of fixing it later is that implementing other consensus rules requires having the right error types, so all later changes would still block on the later PR.

File: zebra-consensus/src/lib.rs [hunk 1]
```diff
@@ -1,13 +1,34 @@
...
+//!    transaction is actually valid in the context of a particular
+//!    chain state.  For instance, a transaction that spends a
+//!    UTXO is only valid if the UTXO remains unspent; a
+//!    shielded transaction spending some note must reveal a nullifier
+//!    not already in the nullifier set, etc.
+//!
+//! *Structural validity* is enforced by the definitions of data
+//! structures in `zebra-chain`.  *Semantic validity* is enforced by the
+//! code in this crate.  *Contextual validity* is enforced in
+//! `zebra-state` when objects are committed to the chain state.
```
Comment (dconnolly) [inline zebra-consensus/src/lib.rs:31]: 💯 💟 🤓

**Review by dconnolly — APPROVED** [review]

**Comment by dconnolly** [comment]
> Hmm, it looks like the error type duplication is pre-existing to this PR: there's `VerifyBlockError` and `VerifyTransactionError` that both shouldn't exist. We could either fix it now, or in a later PR. The downside of fixing it later is that implementing other consensus rules requires having the right error types, so all later changes would still block on the later PR.

I've created https://github.com/ZcashFoundation/zebra/issues/1186 so that we can merge this for now and let the checks one get rebased, and will dedupe the `Error`s separately but v soon.

**Pull request merged**